ygot/render.go

```diff
@@ -174,30 +238,130 @@ func findUpdatedLeaves(leaves map[*path]interface{}, s GoStruct, parentPath []in
+		if err != nil {
+			return nil, fmt.Errorf("can't append path element key: %v", err)
+		}
+		// We copy the elements from the existing elementPath such that when updating
+		// it, then the elements are not modified when the paths are changed.
+		for _, e := range parentPath.stringSlicePath {
+			childPath.stringSlicePath = append(childPath.stringSlicePath, e)
+		}
+		childPath.stringSlicePath = append(childPath.stringSlicePath, keyval)
+		return childPath, nil
+	}
+
+	for _, e := range parentPath.pathElemPath {
+		n := *e
+		childPath.pathElemPath = append(childPath.pathElemPath, &n)
+	}
+
+	return appendgNMIPathElemKey(value, childPath)
+}
+
 // interfacePathAsgNMIPath takes a path that is specified as a slice of empty interfaces
 // and populates a gNMI path message with the string components. It should be noted that this
 // functionality does not comply with the gNMI specification, and should be updated in the
 // future.
-// TODO(robjs): Update this functionality based on adoption of the gNMI structured paths
-// when Pictor is ready to support these.
-func interfacePathAsgNMIPath(p []interface{}) *gnmipb.Path {
+func interfacePathAsgNMIPath(p []interface{}, usePathElem bool) (*gnmipb.Path, error) {
 	pfx := &gnmipb.Path{}
 	for _, e := range p {
-		pfx.Element = append(pfx.Element, fmt.Sprintf("%v", e))
+		if usePathElem {
+			pe, ok := e.(*gnmipb.PathElem)
+			if !ok {
+				return nil, fmt.Errorf("invalid path element %v in path %v, expected gnmipb.PathElem, got: %T", e, p, e)
+			}
+			pfx.Elem = append(pfx.Elem, pe)
+		} else {
+			pfx.Element = append(pfx.Element, fmt.Sprintf("%v", e))
+		}
 	}
-	return pfx
+	return pfx, nil
 }
 
-// stripPrefix removes the specified prefix from the provided path. Returns an error if
-// the prefix is not a valid prefix of path.
-func stripPrefix(path []interface{}, prefix []interface{}) ([]interface{}, error) {
-	for i := range prefix {
-		if path[i] != prefix[i] {
-			return nil, fmt.Errorf("path %v does not have prefix %v", path, prefix)
+// appendgNMIPathElemKey takes an input reflect.Value which must implement KeyHelperGoStruct
+// and appends the keys from it to the last entry in the supplied mapPath, which must be a
+// gNMI PathElem message.
+func appendgNMIPathElemKey(v reflect.Value, p *gnmiPath) (*gnmiPath, error) {
+	if p == nil {
+		return nil, fmt.Errorf("nil path supplied")
+	}
+
+	if !p.isValid() {
+		return nil, fmt.Errorf("invalid structured path in supplied path: %v", p)
+	}
+
+	if p.isStringSlicePath() || !p.isPathElemPath() {
+		return nil, fmt.Errorf("invalid path type to append keys: %v", p)
+	}
+
+	if len(p.pathElemPath) == 0 {
+		return nil, fmt.Errorf("invalid path element path length, can't append keys to 0 length path: %v", p.pathElemPath)
+	}
+
+	np := &gnmiPath{}
+	np.pathElemPath = make([]*gnmipb.PathElem, len(p.pathElemPath))
+	copy(np.pathElemPath, p.pathElemPath)
+	newElem := *np.pathElemPath[len(np.pathElemPath)-1]
+
```

Comment:
Suggestion something like the following to abstract the internals:
np := p.Copy()
newElem := *p.At(p.Len()-1) // or *p.Last() maybe?